lib/HaskellLemmaBucket.thy

```diff
@@ -127,10 +127,14 @@ lemma empty_fail_stateAssert[intro!, simp]:
   "empty_fail (stateAssert P l)"
   unfolding stateAssert_def by simp
 
-lemma haskell_assert_wp:
+lemma haskell_assert_wp[wp]:
```

Comment:
It does lead to a bunch of new `hoare_drop_imps` and `crunch_wps`, but I think I agree with the overall strategy of making sure that automation doesn't lose information. Accidental `haskell_assert_inv` uses are very hard to debug.

Reply:
I should probably have also mentioned that this is what's done in the `rt` branch, and that Corey says there in the PR that made it `[wp]` that it mirrors what's done with the other wp rules like `assert_wp`

Reply:
Yes, I'm happy with the `[wp]` for this one, was just thinking out loud.

Reply:
Is this related to the long-standing annoyance of how in the abstract spec the default wp rule would keep the knowledge of what was asserted, but in the Haskell it'd just get dropped?

Reply:
Yes I think so. I think previously, `haskell_assert_inv` was generated and added to the `[wp]` set by a `crunch` somewhere and was picked up by many proofs, and my guess is that now that `haskell_assert_wp` is added to the `[wp]` set in `Lib`, that rule is now preferred and won't lose information.